zipkin/src/main/java/zipkin/Span.java

```diff
@@ -253,22 +264,13 @@ public Builder merge(Span that) {
         } else if (that.duration != null) {
           this.duration = Math.max(this.duration, that.duration);
         }
-      } else { // duration might need to be recalculated, since we have 2 different timestamps
-        long thisEndTs = this.duration != null ? this.timestamp + this.duration : this.timestamp;
-        long thatEndTs = that.duration != null ? that.timestamp + that.duration : that.timestamp;
-        this.timestamp = Math.min(this.timestamp, that.timestamp);
-        this.duration = Math.max(thisEndTs, thatEndTs) - this.timestamp;
+      } else {
+        // We have 2 different timestamps. Use ApplyTimestampAndDuration to compute final timestamp.
+        this.timestamp = null;
+        this.duration = null;
+        return ApplyTimestampAndDuration.apply(this.build()).toBuilder();
```

Comment:
we can't do this as we only include authoritative data reported to us (as opposed to deriving span's timestamp from annotations). The prior data incorrectly tried to merge authoritatively set timestamps (which was the bug). The better way is to prefer the client as mentioned before.